required/latex-lab/socket-sample.tex

```diff
@@ -0,0 +1,411 @@
+overkill in such a case (or otherwise awkward, because paramerization
+is better done on a higher level instead of individually for small
+blocks of code) and using the template mechanism just to replace one
+block of code with a different one results in a fairly high
+performance hit. It is therefore usually not a good choice.
+
+In theory, it would also be possible to use a hook, but again that is
+basically a misuse of the concept, because in this case there should
+never be more that one block of code inside such a  \enquote{special hook}, so that to alter
+the processing logic one would need to set up rules that replace code
+rather than (as intended in the hook mechanism) execute all code added to a hook.
+
+
+
+For this reason \LaTeX{} now offers a third mechanism: sockets
+which are \enquote{replaceable code blocks} that are referred to by name.
+%
+In a nutshell: instead of having a fixed code block somewhere as part
+of the code that implements a certain programming logic, there is a named socket
+which is assigned a chunk of code to be executed in this place.
+
+This is done by first declaring the socket with:
+\begin{quote}
+\cs{NewSocket}\marg{socket-name}\marg{number-of-arguments}
+\end{quote}
+This is then used in code through
+\begin{quote}
+ \cs{UseSocket}\marg{socket-name}
+\end{quote}
+or, if the socket takes a number of arguments with
+\begin{quote}
+  \cs{UseSocket}\marg{socket-name}\marg{arg\textsubscript{1}}\ldots
+  \marg{arg\textsubscript{number-of-arguments}}
+\end{quote}
+
+In addition, one or several named chunks of code implementing different logic for this
+socket are set up, each with a declaration of the form:
+\begin{quote}
+  \cs{DeclareSocketCode}\marg{socket-name}\marg{chunk-name}\marg{code})
+\end{quote}
+Finally,
+one of them is assigned to the socket:
+\begin{quote}
+\cs{AssignSocketCode}\marg{socket-name}\marg{chunk-name}
+\end{quote}
+If the programming logic should change, then all that is necessary is
+to make a new assignment with \cs{AssignSocketCode} to a different
+\marg{chunk-name}.
+
+If the socket takes arguments, then those need to be provided to
+\cs{UseSocket} and in that case they can be referenced in the \meta{code}
+argument of \cs{DeclareSocketCode} with \verb=#1=, \verb=#2=, etc.
+
+In most cases a named socket is used only in a single place, but there
+is, of course, nothing wrong wth using it in several places, as long
+as the code in all places is supposed to change in the same way if the socket gets reassigned a different chunk of code.
+
+
+\section{Alternative names}
+
+
+By the way:
+\begin{itemize}
+\item
+  should it be \texttt{gassign} ? or some other verb?
+\item
+  And what about \cs{socket\_use:n} ? Should that be \texttt{:nw}
+  as wedon't really control the arguments or should we have
+  \cs{socket\_use:nn}, \cs{socket\_use:nnn}, \ldots, just to make it
+  clearer how many arguments are (supposed to be) picked up?
```

Comment:
In lthooks we used `\hook_use:nw` for the version with arguments, so I think `\socket_use:nw` for consistency (also because the code is the same, regardless of the actual number of arguments).

Reply:
This suggestion @PhelypeOleinik to use `:nw` sounds sensible.

Reply:
This is what I have done in the real code, that file should have been deleted it was an early draft.

Reply:
OK, good.  Sorry about the noise.

Reply:
and it is already the case in the .dtx